### PR TITLE
Add VRL support

### LINKED ISSUE #2077: Add support for VRL
We could probably add support for VRL in the `DocProcessor` for instance.

### PR DESCRIPTION
**Pull request by boraarslan**
Closes #2077

Adds VRL support to transform documents before ingesting.

TODO:
- [x] Add benchmarks
- [x] Fork `vector` repo under `quickwit-oss`

### CODE REVIEW

File: quickwit/quickwit-indexing/benches/doc_process_vrl_bench.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
+use std::sync::Arc;
```
Comment (guilload) [inline quickwit/quickwit-indexing/benches/doc_process_vrl_bench.rs:1]: Where does the bench data come from?

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 2]
```diff
@@ -143,6 +177,83 @@ pub struct DocProcessor {
     timestamp_field_opt: Option<Field>,
     counters: DocProcessorCounters,
     publish_lock: PublishLock,
+    transform_layer: Option<VrlProgram>,
+}
+
+struct VrlProgram {
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:183]: Can we move the `VrlProgram` and related implementations out of `doc_processor`?

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 3]
```diff
@@ -143,6 +177,83 @@ pub struct DocProcessor {
...
+            inner: program,
+            runtime,
+            timezone,
+        })
+    }
+}
+
+impl VrlProgram {
+    pub fn process_doc(&mut self, doc_json: &str) -> Result<String, PrepareDocumentError> {
+        let mut target: ::value::Value =
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:232]:
```suggestion
        let mut target: VrlValue =
```

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 4]
```diff
@@ -143,6 +177,83 @@ pub struct DocProcessor {
...
+        let runtime_result = self
+            .runtime
+            .resolve(&mut target_value, &self.inner, &self.timezone);
+
+        self.runtime.clear();
+
+        match runtime_result {
+            Ok(value) => Ok(value.to_string()),
+            Err(terminate) => {
+                println!("Vrl returned error {}", terminate);
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:252]: Remove.

File: quickwit/quickwit-indexing/src/actors/doc_processor.rs [hunk 5]
```diff
@@ -151,25 +262,43 @@ impl DocProcessor {
...
+        &mut self,
         doc_json: String,
         ctx: &ActorContext<Self>,
     ) -> Result<PreparedDoc, PrepareDocumentError> {
-        // Parse the document
         let _protect_guard = ctx.protect_zone();
+        self._prepare_document(doc_json)
+    }
+
+    pub fn _prepare_document(&mut self, doc_json: String) -> Result<PreparedDoc, PrepareDocumentError> {
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/actors/doc_processor.rs:295]: Oh, I see what you did here. Come on, you know this is ugly. Please, propose a better alternative.

**Review by guilload — COMMENTED** [review]
Looking good.

**Comment by boraarslan** [comment]
At last, CI passes :relieved:

**Review by guilload — APPROVED** [review]

**Pull request merged**